### DISCUSSION TITLE
Firefly III Data Importer: Invalid Client

### DISCUSSION THREAD

**Discussion by Drakon74** [firefly-iii/firefly-iii#5877] [Firefly III]
Hi,

i am using a Synology with Docker installed to run the containers.

My current Setup is as follows:

**Firefly III Container**
- Runs on the NAS under local port 8100
- Is behind Synologys Reverse Proxy and outside port 8001 (so i open it up with https://my-domain:8001 from the internet)

This works flawlessly!

Now i just setup the **Firefly III Data importer**, my setup right now is:
- Runs on the NAS under local port 8101
- Has the Firefly 3 URL Env Variable set to "http://my-nas-ip:8100"
- Has the Vanity URL set to "https://my-domain:8001"
- Is also reachable from the outside over the same reverse proxy under the same domain, but under port 8002 (so: https://my-domain:8002)
- Inside the Firefly 3 Installation i setup a client ID, because i plan that other users can use the same instance, i put in: http://-my-nas-ip:8101/callback as the Callback URL. Also removed the check for the secret stuff.
- The Client ID (a number) i setup as the FIREFLY_III_CLIENT_ID environment variable (just the number)

However this is not working.

When i type in https://my-domain:8002 into my browser it redirects me to https://my-domain:8001 where the Firefly 3 Installation sits. The login mask appears, i log in as normal however, when i logged in i get the following raw text displayed still under the 8001 (so the Firefly 3 Installation) port:

{"error":"invalid_client","error_description":"Client authentication failed","message":"Client authentication failed"}

I set up the Client ID under that very same user account i tried to login too.

I already tried:

- Setting the URLs from IPs to their outside domain ones and vice versa in basically every combination i could think off. Either it doesn't work at all or i get the mentioned error again.

Does someone know what the problem could be? I can't figure it out...

**Comment by JC5** [comment]
Generally speaking there are two addresses. You need the internal one, the one Firefly III can reach over Docker. And the one you browse to. These two can be different. Do you know which ones I mean?

**Reply by Drakon74** [reply]
Hi, i think so. My NAS has the internal IP 192.168.178.60 in my network. The two containers are running on the NAS and have their local ports changed from 8080 to 8100 (Firefly 3 core) and 8101 (firefly 3 data importer).

From my understanding, i have to put in 192.168.178.60:8100 into the ENV Variable FIREFLY_III_URL in the Data Importer. And when creating the Client ID i put in http://192.168.178.60:8101/callback. From my knowledge the containers should be able to communicate this way. 

As VANITY_URL i put in the public address where the Firefly 3 core installation can be reached from the internet.

Am i misunderstanding something here?

**Reply by Drakon74** [reply]
Okay, after some more troubleshooting it appears the reverseproxy is giving me trouble. Trying that very setup from my LAN is working...

**Reply by JC5** [reply]
It looks to be correct, but I'm curious to see what happens in the reverse proxy. Getting the error means that the client ID reaches Firefly III just fine.

**Reply by Drakon74** [reply]
I sadly can't find the error. I tried a bunch of things, different settings and custom header properties. Nothing works. Without the reverse proxy it's working fine, as soon as the reverse proxy comes in, thats it and i get the error from above.

Well, guess i have to do my transactions manually then. Still, thanks for the help.

**Reply by Drakon74** [reply]
So after some more trying and troubleshooting this morning, i made *some* progress... 

It seems i forgot the "TRUSTED_PROXIES" env variable for the data importer, added that. And i changed the callback URL to the external one which goes to the reverse proxy.

Now when i visit the Data importer over the reverse proxy, i get redirected to the firefly 3 core. Log in, i get the messagebox asking if i wanna authorize this App, i click authorize.

It then tries to connect me back to the data importer with /callback with a very long "code" and a smaller "state" variable in the URL but the reverse proxy gives an error back that he can't find the target... any ideas?

The URL and port seems correct... It goes like https://my-domain:8002/callback?code=[Long_code]&state=[State]

**Reply by JC5** [reply]
The URL, the long code and the state are correct. What's the exact error of the reverse proxy?

**Reply by Drakon74** [reply]
Just that it couldn't find the page. It's some pre-made Synology site coming back, basically just saying:

"We're sorry, we couldn't find your requested page." No Error Code sadly.

**Reply by JC5** [reply]
Mmmm. Perhaps the URL is too long, does it work when you cut away parts of the long code?

It won't work of course but it's good debug.

**Reply by Drakon74** [reply]
Yep.. now i come to the error page of the data importer... seems the URL is indeed too long for the reverse proxy i guess...

**Reply by JC5** [reply]
If the reverse proxy generates the error (maybe you have logs?) then it should be a config issue for the reverse proxy. Otherwise it's something for Synology.

**Reply by Drakon74** [reply]
I will check if i can find any logs, but it seems i don't have any access to a config to check. I will probably contact the Synology Support about this. Anyway, thanks for your help!

**Reply by Drakon74** [reply]
It is indeed related to the buffer size and nginx (which is synology using) chocking on the response:

2022/03/12 10:50:51 [error] 2503#2503: *48661 upstream sent too big header while reading response header from upstream

**Reply by Drakon74** [reply]
Yep, increasing the buffer size with the help of this link:

https://www.cyberciti.biz/faq/nginx-upstream-sent-too-big-header-while-reading-response-header-from-upstream/

is indeed fixing the issue. lol.

Sadly that specific config file gets re-written everytime you do -any- change in the reverse proxy settings, so eh... i will try if the import works and if it does, reverse my changes and contact the support about this, what the "right" way would be.
